################################################################################
SHORT DESCRIPTION:
################################################################################
Log to syslog instead of logfiles

################################################################################
LONG DESCRIPTION:
################################################################################
LibXL only has one way to log messages, which uses log files.
In OpenXT, we want all logs to go through syslog.
This patch does that.

################################################################################
CHANGELOG
################################################################################
Authors:
Jed Lejosne <lejosnej@ainfosec.com>

################################################################################
REMOVAL
################################################################################
If syslog support gets added upstream, we should remove this patch

################################################################################
UPSTREAM PLAN
################################################################################
As is it, this patch is not upstream-able.
Givent the opportunity, we could make it right (adding an option to choose files
or syslog) and then upstream it.

################################################################################
INTERNAL DEPENDENCIES
################################################################################

################################################################################
PATCHES
################################################################################
Index: xen-4.6.5/tools/libxl/xl_cmdimpl.c
===================================================================
--- xen-4.6.5.orig/tools/libxl/xl_cmdimpl.c
+++ xen-4.6.5/tools/libxl/xl_cmdimpl.c
@@ -475,7 +475,6 @@ out:
 
 static int do_daemonize(char *name, const char *pidfile)
 {
-    char *fullname;
     pid_t child1;
     int nullfd, ret = 0;
 
@@ -489,19 +488,10 @@ static int do_daemonize(char *name, cons
 
     postfork();
 
-    ret = libxl_create_logfile(ctx, name, &fullname);
-    if (ret) {
-        LOG("failed to open logfile %s: %s",fullname,strerror(errno));
-        exit(-1);
-    }
-
-    CHK_SYSCALL(logfile = open(fullname, O_WRONLY|O_CREAT|O_APPEND, 0644));
-    free(fullname);
+    /* OpenXT: we don't want separate logfiles */
 
     CHK_SYSCALL(nullfd = open("/dev/null", O_RDONLY));
     dup2(nullfd, 0);
-    dup2(logfile, 1);
-    dup2(logfile, 2);
 
     CHK_SYSCALL(daemon(0, 1));
 
Index: xen-4.6.5/tools/libxl/libxl_internal.c
===================================================================
--- xen-4.6.5.orig/tools/libxl/libxl_internal.c
+++ xen-4.6.5/tools/libxl/libxl_internal.c
@@ -199,6 +199,34 @@ char *libxl__dirname(libxl__gc *gc, cons
     return libxl__strndup(gc, s, c - s);
 }
 
+static int xtl_level_to_syslog_level(xentoollog_level x)
+{
+    int s;
+
+    switch(x) {
+    case XTL_DEBUG:
+        s = LOG_DEBUG;
+        break;
+    case XTL_NOTICE:
+        s = LOG_NOTICE;
+        break;
+    case XTL_WARN:
+        s = LOG_WARNING;
+        break;
+    case XTL_ERROR:
+        s = LOG_ERR;
+        break;
+    case XTL_CRITICAL:
+        s = LOG_CRIT;
+        break;
+    default:
+        s = LOG_INFO;
+        break;
+    }
+
+    return s;
+}
+
 void libxl__logv(libxl_ctx *ctx, xentoollog_level msglevel, int errnoval,
              const char *file, int line, const char *func,
              const char *fmt, va_list ap)
@@ -221,10 +249,14 @@ void libxl__logv(libxl_ctx *ctx, xentool
     fileline[sizeof(fileline)-1] = 0;
 
  x:
-    xtl_log(ctx->lg, msglevel, errnoval, "libxl",
-            "%s%s%s%s" "%s",
-            fileline, func&&file?":":"", func?func:"", func||file?": ":"",
-            base);
+    /* xtl_log(ctx->lg, msglevel, errnoval, "libxl", */
+    /*         "%s%s%s%s" "%s", */
+    /*         fileline, func&&file?":":"", func?func:"", func||file?": ":"", */
+    /*         base); */
+    /* OpenXT: we use syslog */
+    syslog(LOG_USER | xtl_level_to_syslog_level(msglevel), "%s%s%s%s%s",
+           fileline, func&&file?":":"", func?func:"", func||file?": ":"", base);
+
     if (base != enomem) free(base);
     errno = esave;
 }
Index: xen-4.6.5/tools/libxl/libxl_internal.h
===================================================================
--- xen-4.6.5.orig/tools/libxl/libxl_internal.h
+++ xen-4.6.5/tools/libxl/libxl_internal.h
@@ -52,6 +52,9 @@
 #include <xenctrl.h>
 #include <xenguest.h>
 
+/* OpenXT: we use syslog */
+#include <syslog.h>
+
 #include "xentoollog.h"
 
 #include <xen/io/xenbus.h>
Index: xen-4.6.5/tools/libxl/xl.c
===================================================================
--- xen-4.6.5.orig/tools/libxl/xl.c
+++ xen-4.6.5/tools/libxl/xl.c
@@ -28,6 +28,9 @@
 #include <inttypes.h>
 #include <regex.h>
 
+/* OpenXT: we use syslog */
+#include <syslog.h>
+
 #include "libxl.h"
 #include "libxl_utils.h"
 #include "libxlutil.h"
@@ -291,6 +294,8 @@ static void xl_ctx_free(void)
         free(lockfile);
         lockfile = NULL;
     }
+    /* OpenXT: we use syslog */
+    closelog();
 }
 
 int main(int argc, char **argv)
@@ -333,6 +338,8 @@ int main(int argc, char **argv)
     logger = xtl_createlogger_stdiostream(stderr, minmsglevel,
         (progress_use_cr ? XTL_STDIOSTREAM_PROGRESS_USE_CR : 0));
     if (!logger) exit(1);
+    /* OpenXT: we use syslog, we could probably comment out the above. */
+    openlog("xl", LOG_CONS, LOG_USER);
 
     atexit(xl_ctx_free);
 
Index: xen-4.6.5/tools/libxl/libxl_dm.c
===================================================================
--- xen-4.6.5.orig/tools/libxl/libxl_dm.c
+++ xen-4.6.5/tools/libxl/libxl_dm.c
@@ -39,25 +39,6 @@ static const char *qemu_xen_path(libxl__
     return QEMU_XEN_PATH;
 }
 
-static int libxl__create_qemu_logfile(libxl__gc *gc, char *name)
-{
-    char *logfile;
-    int rc, logfile_w;
-
-    rc = libxl_create_logfile(CTX, name, &logfile);
-    if (rc) return rc;
-
-    logfile_w = open(logfile, O_WRONLY|O_CREAT|O_APPEND, 0644);
-    free(logfile);
-
-    if (logfile_w < 0) {
-        LOGE(ERROR, "unable to open Qemu logfile");
-        return ERROR_FAIL;
-    }
-
-    return logfile_w;
-}
-
 const char *libxl__domain_device_model(libxl__gc *gc,
                                        const libxl_domain_build_info *info)
 {
@@ -1697,11 +1678,10 @@ void libxl__spawn_local_dm(libxl__egc *e
 
     libxl_ctx *ctx = CTX;
     libxl_domain_config *guest_config = dmss->guest_config;
-    const libxl_domain_create_info *c_info = &guest_config->c_info;
     const libxl_domain_build_info *b_info = &guest_config->b_info;
     const libxl_vnc_info *vnc = libxl__dm_vnc(guest_config);
     char *path;
-    int logfile_w, null;
+    int null;
     int rc;
     char **args, **arg, **envs;
     xs_transaction_t t;
@@ -1758,12 +1738,6 @@ void libxl__spawn_local_dm(libxl__egc *e
         libxl__xs_write(gc, XBT_NULL, libxl__sprintf(gc, "%s/disable_pf", path),
                     "%d", !libxl_defbool_val(b_info->u.hvm.xen_platform_pci));
 
-    logfile_w = libxl__create_qemu_logfile(gc, GCSPRINTF("qemu-dm-%s",
-                                                         c_info->name));
-    if (logfile_w < 0) {
-        rc = logfile_w;
-        goto out;
-    }
     null = open("/dev/null", O_RDONLY);
     if (null < 0) {
         LOGE(ERROR, "unable to open /dev/null");
@@ -1817,14 +1791,13 @@ retry_transaction:
         goto out_close;
     if (!rc) { /* inner child */
         setsid();
-        libxl__exec(gc, null, logfile_w, logfile_w, dm, args, envs);
+        libxl__exec(gc, null, null, null, dm, args, envs);
     }
 
     rc = 0;
 
 out_close:
     if (null >= 0) close(null);
-    if (logfile_w >= 0) close(logfile_w);
 out:
     if (dm_state_fd >= 0) close(dm_state_fd);
     if (rc)
@@ -1893,7 +1866,7 @@ void libxl__spawn_qdisk_backend(libxl__e
     flexarray_t *dm_args, *dm_envs;
     char **args, **envs;
     const char *dm;
-    int logfile_w, null = -1, rc;
+    int null = -1, rc;
     uint32_t domid = dmss->guest_domid;
 
     /* Always use qemu-xen as device model */
@@ -1918,11 +1891,6 @@ void libxl__spawn_qdisk_backend(libxl__e
     libxl__set_qemu_env_for_xsa_180(gc, dm_envs);
     envs = (char **) flexarray_contents(dm_envs);
 
-    logfile_w = libxl__create_qemu_logfile(gc, GCSPRINTF("qdisk-%u", domid));
-    if (logfile_w < 0) {
-        rc = logfile_w;
-        goto error;
-    }
     null = open("/dev/null", O_RDONLY);
     if (null < 0) {
        rc = ERROR_FAIL;
@@ -1955,14 +1923,13 @@ void libxl__spawn_qdisk_backend(libxl__e
         goto error;
     if (!rc) { /* inner child */
         setsid();
-        libxl__exec(gc, null, logfile_w, logfile_w, dm, args, envs);
+        libxl__exec(gc, null, null, null, dm, args, envs);
     }
 
     return;
 
 error:
     assert(rc);
-    if (logfile_w >= 0) close(logfile_w);
     if (null >= 0) close(null);
     dmss->callback(egc, dmss, rc);
     return;
